### ISSUE TITLE
Training is different depending on loss function being decorate with tf.function or not

### ISSUE THREAD

**Issue by borundev** [tensorflow/tensorflow#46923]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.4.1
- Python version: 3.8
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**Describe the current behavior**

I am training with 4 different versions of a class weighted loss function and am getting very different results. These 4 differ only in from_logits and tf.function decorator. I can compare with the result from running model.fit with the standard categorical cross entropy loss function and explicit weights and see which ones are correct and get

1. from_logits = True, no tf.function decorator ---> wrong answer
2. from_logits = True, tf.function decorator ---> wrong answer
3. from_logits = False, no tf.function decorator ---> wrong answer
4. from_logits = False, tf.function decorator ---> correct answer

Incidentally, all the three incorrect answers are equal to each other. This is not related to the numerical stability issue related to use of from_logits as we can see the correct answer actually comes from the one that does not use logits. Furthermore, the standard categorical cross entropy loss function with class weights give the same answer irrespective of the from_logits flag.

**Describe the expected behavior**

All four loss functions should result in the same training.

**Standalone code to reproduce the issue**

https://colab.research.google.com/drive/1LrB6krDIowtzDfmstU-oGi6iHOoDYRYm?usp=sharing

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.3, TF v2.4 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/78e246e675a878b75a9b83f07f06b5e9/46923.ipynb). Thanks!

**Comment by sushreebarsa** [comment]
Was able to replicate the issue in TF 2.6.0-dev20210603,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/8609ad090b93b8a8b33021247cafc6d5/untitled158.ipynb?authuser=1#scrollTo=7mT26aaVjXsf)..Thanks !

**Comment by tilakrayal** [comment]
@borundev,
I was able to execute the code without any issues. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/7ff3766b265c4668f4b0f49bbf45d677/46923.ipynb) and let us know if it resolves the issue. Thank you!
